mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.kt

```diff
@@ -104,36 +104,84 @@ class MockWebServer : ExternalResource(), Closeable {
       Collections.newSetFromMap(ConcurrentHashMap<Socket, Boolean>())
   private val openConnections =
       Collections.newSetFromMap(ConcurrentHashMap<Http2Connection, Boolean>())
-  private val requestCount = AtomicInteger()
-  private var bodyLimit = Long.MAX_VALUE
-  private var serverSocketFactory: ServerSocketFactory? = null
+
+  private val atomicRequestCount = AtomicInteger()
+
+  /**
+   * The number of HTTP requests received thus far by this server. This may exceed the number of
+   * HTTP connections when connection reuse is in practice.
+   */
+  val requestCount: Int
+    get() = atomicRequestCount.get()
+
+  /** The number of bytes of the POST body to keep in memory to the given limit. */
+  var bodyLimit = Long.MAX_VALUE
+
+  var serverSocketFactory: ServerSocketFactory? = null
+    set(value) {
+      check(executor == null) { "setServerSocketFactory() must be called before start()" }
+      field = value
+    }
+
   private var serverSocket: ServerSocket? = null
   private var sslSocketFactory: SSLSocketFactory? = null
   private var executor: ExecutorService? = null
   private var tunnelProxy: Boolean = false
   private var clientAuth = CLIENT_AUTH_NONE
+
   /**
-   * Returns the dispatcher used to respond to HTTP requests.
-   * The default dispatcher is a [QueueDispatcher] but other dispatchers can be configured.
+   * The dispatcher used to respond to HTTP requests. The default dispatcher is a [QueueDispatcher],
+   * which serves a fixed sequence of responses from a [queue][enqueue].
    *
-   * Sets the dispatcher used to match incoming requests to mock responses. The default dispatcher
-   * simply serves a fixed sequence of responses from a [queue][enqueue]; custom
-   * dispatchers can vary the response based on timing or the content of the request.
+   * Other dispatchers can be configured. They can vary the response based on timing or the content
+   * of the request.
    */
   var dispatcher: Dispatcher = QueueDispatcher()
-  private var port = -1
-  private var inetSocketAddress: InetSocketAddress? = null
-  private var protocolNegotiationEnabled = true
-  private var protocols = immutableListOf(Protocol.HTTP_2, Protocol.HTTP_1_1)
 
-  private var started: Boolean = false
+  var port: Int = -1
+    get() {
+      before()
```

Comment:
this is my least favorite change. I don’t like vals that have side effects when you read them.

This is consistent with hostName below. I’m tempted to change both methods to only be available through a function rather than a val.

Reply:
It's tough since this was the existing 3.14 behaviour also.  Essentially for any awkward kotlin source changes, it only get's harder to make these changes in 4.1.  So make them now since you already have made a few.

Reply:
I’m gonna keep it as-is for now.